src/tactic/linear_combination.lean

```diff
@@ -0,0 +1,386 @@
+  <|> tactic.fail ("The type of the left and right sides of the goal " ++
+    "must fulfill the 'add_group' condition in order to match the linear " ++
+    "combination to the target.")
+
+
+
+/-! ### Part 3: Matching the Linear Combination to the Target -/
+
+
+/--
+This tactic changes the goal to be that the lefthand side of the target is
+  equal to the lefthand side of the given expression.  For example,
+  if `hsum_on_left` is 5*x - 5*y = 0, and the target is -5*y + 5*x = 0, this
+  tactic will change the target to be -5*y + 5*x = 5*x - 5*y.
+
+This tactic only should be used when the target is an equality whose right side
+  is 0.
+
+* Input:
+  * `hsum_on_left` : expr, which should be an equality whose right side is 0
+
+* Output: tactic unit
+-/
+meta def set_goal_to_hleft_eq_tleft (hsum_on_left : expr) : tactic unit :=
+do
+  { change_goal ← tactic.to_expr ``(replace_eq_expr %%hsum_on_left),
+    tactic.apply change_goal,
+    pure () }
+  <|> tactic.fail ("The type of the left and right sides of each equality " ++
+    "must fulfill the 'has_zero' condition.")
+
+/--
+This tactic attempts to prove the goal by normalizing the target if the
+`normalize` field of the given configuration is true.
+
+* Input:
+  * `config` : a linear_combination_config, which determines the tactic used
+      for normalization if normalization is done
+
+* Outout: tactic unit
+-/
+meta def prove_equal_if_desired (config : linear_combination_config) :
+  tactic unit :=
+when config.normalize config.normalization_tactic
+
+/-! ### Part 4: Completed Tactic -/
+
+
+/--
+This is a tactic that attempts to prove the target by creating and applying a
+  linear combination of a list of equalities.  (If the `normalize` field of the
+  configuration is set to ff, then the tactic will simply set the user up to
+  prove their target using the linear combination instead of attempting to
+  finish the proof.)
+
+Note: The left and right sides of all the equations should have the same
+  ring type, and the coefficients should also have this type.  This type must
+  have the has_mul and add_group properties.  Also note that the target must
+  involve at least one variable.
+
+* Input:
+  * `heqs` : a list of names, referring to equations in the local context
+  * `coeffs` : a list of coefficients to be multiplied with the corresponding
+    equations in the list of names
+  * `config` : a linear_combination_config, which determines the tactic used
+      for normalization; by default, this value is the standard configuration
+      for a linear_combination_config
+
+* Output: tactic unit
+-/
+meta def linear_combination (heqs : list name) (coeffs : list pexpr)
+  (config : linear_combination_config := {}) : tactic unit :=
+do
+  hsum ← make_sum_of_hyps heqs coeffs,
+  hsum_on_left ← move_to_left_side hsum,
+  move_target_to_left_side,
+  set_goal_to_hleft_eq_tleft hsum_on_left,
+  prove_equal_if_desired config
```

Comment:
Actually, I don't know if @agoldb10 had this in mind, but having things factored out like this could be helpful if/when the tactic gets extended to modules. I wouldn't change it!